### PR TITLE
test: add performance benchmarks for the phase0 block attributes

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Observe the performance regressions.

**Description**

Add a base performance benchmark for phase0 block attributes. 

Added new metric and a graph, which will helps us to identify any issue with the individual processing steps. 

image

part of #5793

**Steps to test or reproduce**

Run all tests.

### CODE REVIEW

File: packages/beacon-node/src/chain/opPools/opPool.ts [hunk 1]
```diff
@@ -196,12 +196,12 @@ export class OpPool {
       const slashableIndices = new Set<ValidatorIndex>();
       for (let i = 0; i < attesterSlashing.intersectingIndices.length; i++) {
         const index = attesterSlashing.intersectingIndices[i];
-        const validator = state.validators.getReadonly(index);
```
Comment (nazarhussain) [inline packages/beacon-node/src/chain/opPools/opPool.ts:200]: Initialize validator object when needed.

File: packages/beacon-node/src/chain/opPools/opPool.ts [hunk 2]
```diff
@@ -237,7 +254,13 @@ export class OpPool {
         }
       }
     }
+    endVoluntaryExits &&
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/opPools/opPool.ts]: I think this (and others) can be written `endVoluntaryExits?.(...)`

File: packages/beacon-node/src/metrics/metrics/beacon.ts [hunk 3]
```diff
@@ -121,6 +121,33 @@ export function createBeaconMetrics(register: RegistryMetricCreator) {
       buckets: [0.1, 1, 2, 4, 10],
       labelNames: ["source"],
     }),
+    blockProductionTimeSteps: register.histogram<"step" | "source">({
```
Comment (twoeths) [inline packages/beacon-node/src/metrics/metrics/beacon.ts]: from now on we don't want to track more than one label per metric due to its performance issue, see https://github.com/ChainSafe/js-libp2p-gossipsub/pull/433

File: packages/beacon-node/test/perf/chain/opPools/opPool.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+import {generateBlsToExecutionChanges} from "../../../fixtures/capella.js";
+import {
+  generateIndexedAttestations,
+  generateSignedBeaconBlockHeader,
+  generateVoluntaryExits,
+} from "../../../fixtures/phase0.js";
+import {BlockType} from "../../../../src/chain/interface.js";
+
+// Aug 11 2021
+// getAttestationsForBlock
```
Comment (twoeths) [inline packages/beacon-node/test/perf/chain/opPools/opPool.test.ts]: remove this as it does not belong to tests in this file

File: packages/beacon-node/test/perf/chain/opPools/opPool.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      const pool = new OpPool();
+      fillAttesterSlashing(pool, originalState, MAX_ATTESTER_SLASHINGS);
+      fillProposerSlashing(pool, originalState, MAX_PROPOSER_SLASHINGS);
+      fillVoluntaryExits(pool, originalState, MAX_VOLUNTARY_EXITS);
+      fillBlsToExecutionChanges(pool, originalState, MAX_BLS_TO_EXECUTION_CHANGES);
+
+      return pool;
+    },
+    fn: (pool) => {
+      pool.getSlashingsAndExits(originalState, BlockType.Full, null);
```
Comment (twoeths) [inline packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:42]: if this does not modify oppool and the time tracked is too small, consider using `runsFactor` to make it more precise

Comment (nazarhussain) [reply packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:42]: It's less than 100ms which is fine to measure.

File: packages/beacon-node/src/metrics/metrics/beacon.ts [hunk 6]
```diff
@@ -121,6 +121,38 @@ export function createBeaconMetrics(register: RegistryMetricCreator) {
       buckets: [0.1, 1, 2, 4, 10],
       labelNames: ["source"],
     }),
+    executionBlockProductionTimeSteps: register.histogram<"step">({
+      name: "beacon_block_production_execution_steps_seconds",
+      help: "Detailed steps runtime of execution block production",
+      buckets: [0.1, 1, 2, 4, 10],
```
Comment (twoeths) [inline packages/beacon-node/src/metrics/metrics/beacon.ts]: buckets value seems to big for each step. If a step takes more than 1s then it's already an issue so I suggest removing `2, 4, 10` and track more `< 1` buckets

File: packages/beacon-node/src/metrics/metrics/beacon.ts [hunk 7]
```diff
@@ -121,6 +121,38 @@ export function createBeaconMetrics(register: RegistryMetricCreator) {
...
+       * - eth1DataAndDeposits
+       * - syncAggregate
+       * - executionPayload
+       */
+      labelNames: ["step"],
+    }),
+    builderBlockProductionTimeSteps: register.histogram<"step">({
+      name: "beacon_block_production_builder_steps_seconds",
+      help: "Detailed steps runtime of builder block production",
+      buckets: [0.1, 1, 2, 4, 10],
```
Comment (twoeths) [inline packages/beacon-node/src/metrics/metrics/beacon.ts]: same to above comment

File: packages/beacon-node/test/perf/chain/opPools/opPool.test.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    });
+  }
+
+  return pool;
+}
+
+function fillProposerSlashing(pool: OpPool, state: CachedBeaconStateAltair, count: number): OpPool {
+  for (const blockHeader of generateSignedBeaconBlockHeader(state, count)) {
+    pool.insertProposerSlashing({
+      signedHeader1: ssz.phase0.SignedBeaconBlockHeaderBigint.fromJson(
```
Comment (twoeths) [inline packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:79]: why do you have to do `fromJson/toJson` here? if you don't want to mutate the opPool then there should be `clone()` method that's more straightforward

File: packages/beacon-node/test/perf/chain/opPools/opPool.test.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    },
+    fn: (pool) => {
+      pool.getSlashingsAndExits(originalState, BlockType.Full, null);
+    },
+  });
+});
+
+function fillAttesterSlashing(pool: OpPool, state: CachedBeaconStateAltair, count: number): OpPool {
+  for (const attestation of generateIndexedAttestations(state, count)) {
+    pool.insertAttesterSlashing({
```
Comment (twoeths) [inline packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:67]: need to look into the logic of processing an `AttesterSlashing`, I don't think it's a valid `AttesterSlashing` if both attestations are the same

also check the `fromJson/toJson` comment below

Comment (twoeths) [reply packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:67]: if this is only for the `produceBlockBody` and it does not depend on the validation of `AttesterSlashing`, maybe just drop a comment explaining it

File: packages/beacon-node/test/perf/chain/opPools/opPool.test.ts (hunk 8 above)
Comment (nazarhussain) [reply packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:79]: First object is `SignedBeaconBlockHeader` while the `insertProposerSlashing` accepts  `SignedBeaconBlockHeaderBigint`, so why have to convert it via `fromJson/toJson`

File: packages/beacon-node/test/perf/chain/opPools/opPool.test.ts (hunk 9 above)
Comment (nazarhussain) [reply packages/beacon-node/test/perf/chain/opPools/opPool.test.ts:67]: The `opPool` is not checking the validity of the slashings, so for performance tests it does not matter.

File: packages/beacon-node/src/metrics/metrics/beacon.ts (hunk 6 above)
Comment (nazarhussain) [reply packages/beacon-node/src/metrics/metrics/beacon.ts]: I removed the buckets completely, as these are not intended in the metric graph we need.

**Review by twoeths — APPROVED** [review]

**Comment by nazarhussain** [comment]
As the failed perf tests are not related to the code changes in this PR, so why merging it.

File: packages/beacon-node/src/metrics/metrics/beacon.ts [hunk 10]
```diff
@@ -121,6 +121,38 @@ export function createBeaconMetrics(register: RegistryMetricCreator) {
       buckets: [0.1, 1, 2, 4, 10],
       labelNames: ["source"],
     }),
+    executionBlockProductionTimeSteps: register.histogram<"step">({
+      name: "beacon_block_production_execution_steps_seconds",
+      help: "Detailed steps runtime of execution block production",
+      buckets: [0.01, 0.1, 0.2, 0.5, 1],
+      /**
```
Comment (nflaig) [inline packages/beacon-node/src/metrics/metrics/beacon.ts:128]: I feel like commenting those values here is not the right approach. We should use an enum in such cases as we did for block production source.

Might be worth to look into how we could make label values type safe.

**Comment by nflaig** [comment]
> As the failed perf tests are not related to the code changes in this PR, so why merging it.

@nazarhussain are you sure about that? it failed with same errors again on unstable benchmark run https://github.com/ChainSafe/lodestar/actions/runs/7103589338/job/19336631592 while before it was consistently passing.

File: packages/beacon-node/src/metrics/metrics/beacon.ts (hunk 10 above)
Comment (nazarhussain) [reply packages/beacon-node/src/metrics/metrics/beacon.ts:128]: We have to change the types from pom-client for that. That's a bigger task to do. Feel free to create a separate issue to track in future.

**Comment by nazarhussain** [comment]
> @nazarhussain are you sure about that? it failed with same errors again on unstable benchmark run

Some of 'state-transition' perf tests were failing earlier. And this PR didn't changed any code for that package. 

Will look on unstable branch tests in some time.

**Comment by dapplion** [comment]
> As the failed perf tests are not related to the code changes in this PR, so why merging it.

Are you sure? Please investigate

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.13.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.13.0) :tada:

**Pull request merged**